core/src/main/java/org/infinispan/cache/impl/SimpleLocalCacheImpl.java

```diff
@@ -0,0 +1,1608 @@
+		for (Object key : keys) {
+			assertKeyNotNull(key);
+			InternalCacheEntry<K, V> entry = getDataContainer().get(key);
+			if (entry != null) {
+				V value = entry.getValue();
+				if (hasListeners) {
+					cacheNotifier.notifyCacheEntryVisited((K) key, value, true, ImmutableContext.INSTANCE, null);
+					cacheNotifier.notifyCacheEntryVisited((K) key, value, false, ImmutableContext.INSTANCE, null);
+				}
+				map.put(entry.getKey(), entry);
+			}
+		}
+		return map;
+	}
+
+	@Override
+	public EntryIterable<K, V> filterEntries(KeyValueFilter<? super K, ? super V> filter) {
+		return new FilteredEntryIterable(filter);
+	}
+
+	@Override
+	public Map<K, V> getGroup(String groupName) {
+		return Collections.EMPTY_MAP;
+	}
+
+	@Override
+	public void removeGroup(String groupName) {
+	}
+
+	@Override
+	public AvailabilityMode getAvailability() {
+		return AvailabilityMode.AVAILABLE;
+	}
+
+	@Override
+	public void setAvailability(AvailabilityMode availabilityMode) {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public void evict(K key) {
+		ByRef<InternalCacheEntry<K, V>> oldEntryRef = new ByRef<>(null);
+		getDataContainer().compute(key, (k, oldEntry, factory) -> {
+			if (!isNull(oldEntry)) {
+				oldEntryRef.set(oldEntry);
+			}
+			return null;
+		});
+		InternalCacheEntry<K, V> oldEntry = oldEntryRef.get();
+		if (hasListeners && oldEntry != null) {
+			cacheNotifier.notifyCacheEntryEvicted(oldEntry.getKey(), oldEntry.getValue(), ImmutableContext.INSTANCE, null);
+		}
+	}
+
+	@Override
+	public Configuration getCacheConfiguration() {
+		return configuration;
+	}
+
+	@Override
+	public EmbeddedCacheManager getCacheManager() {
+		return cacheManager;
+	}
+
+	@Override
+	public AdvancedCache<K, V> getAdvancedCache() {
+		return this;
+	}
+
+	@Override
+	public ComponentStatus getStatus() {
+		return componentRegistry.getStatus();
+	}
+
+	@ManagedAttribute(
+			description = "Returns the cache status",
+			displayName = "Cache status",
+			dataType = DataType.TRAIT,
+			displayType = DisplayType.SUMMARY
+	)
+	public String getCacheStatus() {
+		return getStatus().toString();
+	}
+
+	@Override
+	public int size() {
+		// we have to iterate in order to provide precise result in case of expiration
+		long now = Long.MIN_VALUE;
+		int size = 0;
+		DataContainer<K, V> dataContainer = getDataContainer();
+		for (InternalCacheEntry<K, V> entry : dataContainer) {
+			if (entry.canExpire()) {
+				if (now == Long.MIN_VALUE) now = timeService.wallClockTime();
+				if (entry.isExpired(now)) {
+					final long NOW = now;
+					dataContainer.compute(entry.getKey(), (key, oldEntry, factory) -> {
+						if (entry.isExpired(NOW)) return null;
+						return oldEntry;
+					});
+				} else {
+					++size;
+				}
+			} else {
+				++size;
+			}
+		}
+		return size;
+	}
+
+	@Override
+	public boolean isEmpty() {
+		return getDataContainer().size() == 0;
```

Comment:
This would be more performant if we used the iterator and found the first non expired entry.